src/main/java/io/supertokens/bulkimport/BulkImportUserUtils.java

```diff
@@ -0,0 +1,167 @@
+            }
+
+            if(deviceName != null) {
+                deviceName = deviceName.trim();
+            }
+            totpDevices.add(new TotpDevice(secretKey, period.intValue(), skew.intValue(), deviceName));
+        }
+        return totpDevices;
+    }
+
+    private static List<LoginMethod> getParsedLoginMethods(JsonObject userData, List<String> errors) {
+      JsonArray jsonLoginMethods = parseAndValidateField(userData, "loginMethods", ValueType.ARRAY_OF_OBJECT, true, JsonArray.class, errors, ".");
+
+      if (jsonLoginMethods == null) {
+          return new ArrayList<>();
+      }
+
+      if (jsonLoginMethods.size() == 0) {
+          errors.add("At least one loginMethod is required.");
+          return new ArrayList<>();
+      }
+
+      Boolean hasPrimaryLoginMethod = false;
+
+      List<LoginMethod> loginMethods = new ArrayList<>();
+      for (JsonElement jsonLoginMethod : jsonLoginMethods) {
+          JsonObject jsonLoginMethodObj = jsonLoginMethod.getAsJsonObject();
+
+          if (validateJsonFieldType(jsonLoginMethodObj, "isPrimary", ValueType.BOOLEAN)) {
+              if (jsonLoginMethodObj.get("isPrimary").getAsBoolean()) {
+                  if (hasPrimaryLoginMethod) {
+                      errors.add("No two loginMethods can have isPrimary as true.");
+                  }
+                  hasPrimaryLoginMethod = true;
+              }
+          }
+
+          String recipeId = parseAndValidateField(jsonLoginMethodObj, "recipeId", ValueType.STRING, true, String.class, errors, " for a loginMethod.");
+          String tenantId = parseAndValidateField(jsonLoginMethodObj, "tenantId", ValueType.STRING, false, String.class, errors, " for a loginMethod.");
+          Boolean isVerified = parseAndValidateField(jsonLoginMethodObj, "isVerified", ValueType.BOOLEAN, false, Boolean.class, errors, " for a loginMethod.");
+          Boolean isPrimary = parseAndValidateField(jsonLoginMethodObj, "isPrimary", ValueType.BOOLEAN, false, Boolean.class, errors, " for a loginMethod.");
+          Number timeJoined = parseAndValidateField(jsonLoginMethodObj, "timeJoinedInMSSinceEpoch", ValueType.NUMBER, false, Number.class, errors, " for a loginMethod");
+          Long timeJoinedInMSSinceEpoch = timeJoined != null ? timeJoined.longValue() : 0;
+
+          if ("emailpassword".equals(recipeId)) {
+            String email = parseAndValidateField(jsonLoginMethodObj, "email", ValueType.STRING, true, String.class, errors, " for an emailpassword recipe.");
+            String passwordHash = parseAndValidateField(jsonLoginMethodObj, "passwordHash", ValueType.STRING, true, String.class, errors, " for an emailpassword recipe.");
+            String hashingAlgorithm = parseAndValidateField(jsonLoginMethodObj, "hashingAlgorithm", ValueType.STRING, true, String.class, errors, " for an emailpassword recipe.");
+
+            email = email != null ? Utils.normaliseEmail(email) : null;
```

Comment:
why you checking for email != null if it must be provided?

Reply:
same goes for hashingAlgorithm

Reply:
> why you checking for email != null if it must be provided?

To ensure we return an exhaustive list of errors to the user, we continue parsing the JSON object even after we find required fields missing.  In this case, we will execute the line 138 even if email will be null. This is why we need to explicitly handle this case.